Client/Source/Client.cpp

```diff
@@ -0,0 +1,361 @@
+#include <cassert>
+#include <thread>
+
+#include "Client.h"
+
+
+namespace vasily
+{
+
+inline const config::Config<std::string, std::string, std::string_view, int, int, long long>
+    Client::CONFIG
+{
+    { "in.txt" },
+    { "out.txt" },
+    { "172.27.221.60", 14 },
+    59003,
+    59002,
+    1000
+};
+ 
+Client::Client(const int layerPort, const std::string_view serverIP, const WorkMode workMode,
+               QObject* parent)
+    : QObject(parent),
+      _layerPort(layerPort),
+      _socketForLayer(std::make_unique<QTcpSocket>(this)),
+      _serverReceivingPort(0),
+      _receivingSocket(nullptr),
+      _serverSendingPort(0),
+      _sendingSocket(nullptr),
+      _serverIP(serverIP),
+      _start(std::chrono::steady_clock::now()),
+      _workMode(workMode),
+      _logger(CONFIG.get<CAST(Param::DEFAULT_IN_FILE_NAME)>(),
+              CONFIG.get<CAST(Param::DEFAULT_OUT_FILE_NAME)>())
+{
+    _printer.writeLine(std::cout, "Layer Port:", layerPort, "Layer IP:", serverIP);
+
+    connect(_socketForLayer.get(), &QTcpSocket::readyRead, this, &Client::slotReadFromLayer);
+
+    connect(_socketForLayer.get(), &QTcpSocket::disconnected, this, &Client::slotServerDisconnected,
+            Qt::QueuedConnection);
+
+    connect(this, &Client::signalToSend, this, &Client::slotSendDataToLayer);
+}
+
+Client::Client(const int serverReceivingPort, const int serverSendingPort,
+               const std::string_view serverIP, const WorkMode workMode, QObject* parent)
+    : QObject(parent),
+      _layerPort(0),
+      _socketForLayer(nullptr),
+      _serverReceivingPort(serverReceivingPort),
+      _receivingSocket(std::make_unique<QTcpSocket>(this)),
+      _serverSendingPort(serverSendingPort),
+      _sendingSocket(std::make_unique<QTcpSocket>(this)),
+      _serverIP(serverIP),
+      _start(std::chrono::steady_clock::now()),
+      _workMode(workMode),
+      _logger(CONFIG.get<CAST(Param::DEFAULT_IN_FILE_NAME)>(),
+              CONFIG.get<CAST(Param::DEFAULT_OUT_FILE_NAME)>())
+{
+    _printer.writeLine(std::cout, "Server Receiving Port:", serverReceivingPort,
+                       "Server Sending Port:", serverSendingPort, "Server IP:", serverIP);
+
+    connect(_receivingSocket.get(), &QTcpSocket::readyRead, this, &Client::slotReadFromServer);
+    connect(_receivingSocket.get(), &QTcpSocket::disconnected, this,
+            &Client::slotServerDisconnected, Qt::QueuedConnection);
+
+    ///connect(_sendingSocket.get(), &QTcpSocket::readyRead, this, &Client::slotClientRead);
+    connect(_sendingSocket.get(), &QTcpSocket::disconnected, this, &Client::slotServerDisconnected,
+            Qt::QueuedConnection);
```

Comment:
Do you need 68 line?

Reply:
Yes, I need this for debug. If I confuse ports, I add a signal to the send socket and check if the data would be received.